// @ngneat/transloco/esm5/lib/template-handler.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/template-handler.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { TemplateRef, ComponentFactoryResolver } from '@angular/core';
import { isString } from './helpers';
import { TranslocoLoaderComponent } from './loader-component.component';
var TemplateHandler = /** @class */ (function () {
    function TemplateHandler(view, vcr) {
        this.view = view;
        this.vcr = vcr;
        this.injector = this.vcr.injector;
    }
    /**
     * @return {?}
     */
    TemplateHandler.prototype.attachView = /**
     * @return {?}
     */
    function () {
        if (this.view instanceof TemplateRef) {
            this.vcr.createEmbeddedView(this.view);
        }
        else if (isString(this.view)) {
            /** @type {?} */
            var componentRef = this.createComponent(TranslocoLoaderComponent);
            componentRef.instance.html = this.view;
            componentRef.hostView.detectChanges();
        }
        else {
            this.createComponent(this.view);
        }
    };
    /**
     * @return {?}
     */
    TemplateHandler.prototype.detachView = /**
     * @return {?}
     */
    function () {
        this.vcr.clear();
    };
    /**
     * @private
     * @template T
     * @param {?} cmp
     * @return {?}
     */
    TemplateHandler.prototype.createComponent = /**
     * @private
     * @template T
     * @param {?} cmp
     * @return {?}
     */
    function (cmp) {
        /** @type {?} */
        var cfr = this.injector.get(ComponentFactoryResolver);
        /** @type {?} */
        var factory = cfr.resolveComponentFactory(cmp);
        return this.vcr.createComponent(factory);
    };
    return TemplateHandler;
}());
export { TemplateHandler };
if (false) {
    /**
     * @type {?}
     * @private
     */
    TemplateHandler.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    TemplateHandler.prototype.view;
    /**
     * @type {?}
     * @private
     */
    TemplateHandler.prototype.vcr;
}
